### PR TITLE
Add business logic for all patients in facility view

### PR DESCRIPTION
**Pull request by vinaysshenoy**
Pivotal Story Link: https://www.pivotaltracker.com/story/show/166784275

### CODE REVIEW

File: app/src/test/java/org/simple/clinic/allpatientsinfacility/AllPatientsInFacilityUiChangeProducerTest.kt [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+
+    // then
+    verifyZeroInteractions(view)
+  }
+
+  @Test
+  fun `when a facility is fetched, then do nothing`() {
+    // given
+    val fetchingPatientsState = AllPatientsInFacilityViewState
+        .FETCHING_PATIENTS_STATE
```
Comment (sanchita) [inline app/src/test/java/org/simple/clinic/allpatientsinfacility/AllPatientsInFacilityUiChangeProducerTest.kt]: I think `AllPatientsInFacilityViewState.FETCHING_PATIENTS_STATE` can be extracted to a variable which is an initial state. 
Or we could rename `FETCHING_PATIENTS_STATE` to `FETCHING_FACILITIES`

File: doc/arch/new-arch-wip.md [hunk 2]
```diff
@@ -0,0 +1,39 @@
+# Refactoring : In-Progress Towards Mobius
+**Note:** This is a WIP doc until we figure out what the new arch finally looks like, at what point it will become an ADR.
```
[the remark this answers is not in the document]
Comment (vinaysshenoy) [reply doc/arch/new-arch-wip.md]: What I meant to say was "When we figure out what the intermediate arch looks like, it will become an ADR." I will reword this to make it clearer.

File: app/src/main/java/org/simple/clinic/allpatientsinfacility/AllPatientsInFacilityViewState.kt [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+import org.simple.clinic.facility.Facility
+import org.simple.clinic.patient.PatientSearchResult
+
+data class AllPatientsInFacilityViewState(
+    val patientsQueried: Boolean,
+    val facility: Facility? = null,
+    val patients: List<PatientSearchResult> = emptyList()
+) {
+  companion object {
+    val FETCHING_PATIENTS_STATE = AllPatientsInFacilityViewState(false)
```
[the remark this answers is not in the document]
Comment (vinaysshenoy) [reply app/src/main/java/org/simple/clinic/allpatientsinfacility/AllPatientsInFacilityViewState.kt]: Yes, this is the initial state of the screen. By convention, all view state objects will have one public state which will be the initial state of that screen.

File: app/src/main/java/org/simple/clinic/allpatientsinfacility/AllPatientsInFacilityUiChangeProducer.kt [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+  override fun apply(viewStates: Observable<AllPatientsInFacilityViewState>): ObservableSource<AllPatientsInFacilityUiChange> {
+    val queriedPatientsViewStates = viewStates.filter { it.patientsQueried }
+
+    val noPatientsInFacilityUiChanges = queriedPatientsViewStates
+        .filter { it.patients.isEmpty() }
+        .map { { view: AllPatientsInFacilityView -> view.showNoPatientsFound(it.facility!!) } }
+
+    val hasPatientsInFacilityUiChanges = queriedPatientsViewStates
+        .filter { it.patients.isNotEmpty() }
+        .map { { view: AllPatientsInFacilityView -> view.showPatients(it.facility!!, it.patients) } }
```
[the remark this answers is not in the document]
Comment (vinaysshenoy) [reply app/src/main/java/org/simple/clinic/allpatientsinfacility/AllPatientsInFacilityUiChangeProducer.kt]: Duplicate comment.

File: app/src/main/java/org/simple/clinic/allpatientsinfacility/AllPatientsInFacilityViewState.kt (hunk 3 above)
[the remark this answers is not in the document]
Comment (vinaysshenoy) [reply app/src/main/java/org/simple/clinic/allpatientsinfacility/AllPatientsInFacilityViewState.kt]: This will be more evident in a future PR where this state is provided externally.

File: app/src/main/java/org/simple/clinic/allpatientsinfacility/AllPatientsInFacilityUiChangeProducer.kt [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+
+typealias AllPatientsInFacilityUiChange = (AllPatientsInFacilityView) -> Unit
+
+class AllPatientsInFacilityUiChangeProducer : ObservableTransformer<AllPatientsInFacilityViewState, AllPatientsInFacilityUiChange> {
+  override fun apply(viewStates: Observable<AllPatientsInFacilityViewState>): ObservableSource<AllPatientsInFacilityUiChange> {
+    val queriedPatientsViewStates = viewStates.filter { it.patientsQueried }
+
+    val noPatientsInFacilityUiChanges = queriedPatientsViewStates
+        .filter { it.patients.isEmpty() }
+        .map { { view: AllPatientsInFacilityView -> view.showNoPatientsFound(it.facility!!) } }
```
[the remark this answers is not in the document]
Comment (vinaysshenoy) [reply app/src/main/java/org/simple/clinic/allpatientsinfacility/AllPatientsInFacilityUiChangeProducer.kt]: Done.

File: app/src/test/java/org/simple/clinic/allpatientsinfacility/AllPatientsInFacilityUiChangeProducerTest.kt (hunk 1 above)
Comment (vinaysshenoy) [reply app/src/test/java/org/simple/clinic/allpatientsinfacility/AllPatientsInFacilityUiChangeProducerTest.kt]: Good point. Fixed.

File: app/src/main/java/org/simple/clinic/allpatientsinfacility/AllPatientsInFacilityViewStateProducer.kt [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+
+                val facilityViewStates = sharedFacilities
+                    .map { initialState.facilityFetched(it) }
+
+                val patientsViewStates = sharedFacilities
+                    .switchMap { patientRepository.allPatientsInFacility(it) }
+                    .withLatestFrom(facilityViewStates)
+                    .map { (searchResults, state) ->
+                      if (searchResults.isEmpty()) state.noPatients() else state.hasPatients(searchResults)
+                    }
```
[the remark this answers is not in the document]
Comment (vinaysshenoy) [reply app/src/main/java/org/simple/clinic/allpatientsinfacility/AllPatientsInFacilityViewStateProducer.kt]: Done.
